types/src/system/auction/detail.rs

```diff
@@ -143,7 +143,23 @@ pub fn get_seigniorage_recipients_snapshot<P>(
 where
     P: StorageProvider + RuntimeProvider + ?Sized,
 {
-    Ok(read_from(provider, SEIGNIORAGE_RECIPIENTS_SNAPSHOT_KEY)?)
+    let era_keys = provider.get_keys(&KeyTag::EraValidators)?;
+
+    let mut ret = BTreeMap::new();
+
+    for key in era_keys {
+        let era_id = match key {
+            Key::EraValidators(era_id) => era_id,
+            _ => return Err(Error::InvalidKeyVariant),
+        };
+        let recipient = match provider.read_era_validators(era_id)? {
+            Some(recipient) => recipient,
+            None => return Err(Error::ValidatorNotFound),
+        };
+        ret.insert(era_id, recipient);
+    }
+
+    Ok(ret)
```

Comment:
This function returns all seigniorage recipients, and `set_seigniorage_recipients_snapshot` writes all passed recipients. 
This might have some side effects here:
https://github.com/darthsiroftardis/casper-node/blob/117e71b3c81c017f9b83c816069297d90b4b066a/types/src/system/auction/mod.rs#L425-L444. First it gets all the seigniorage recipients, then trims it to only have N entries, and then it writes snapshot back. 

I'm thinking it will be more efficient if we'd read only a fixed amount of seigniorage recipients (directly through read_era_validators, not iterating through keys) limited by the snapshot size?

Reply:
@henrytill Thoughts??

Reply:
They key piece of @mpapierski's critique is correct: we need to limit the number of keys we read so that we aren't returning _all_ the eras, just the "window" that was there before.  The window's size is essentially the auction delay IIRC.

Reply:
> Edit: I think I take most of it back. We might not need to get the seigniorage snapshot amnd set the seigniorage snapshot there at all. Perhaps just write "recipients" map under Key::EraValidators(delayed_era)?

@mpapierski, not sure I fully grasp your thinking here, can you elaborate?

Reply:
https://github.com/darthsiroftardis/casper-node/blob/117e71b3c81c017f9b83c816069297d90b4b066a/types/src/system/auction/mod.rs#L425-L444 this code could probable become just:

```rust
  let delayed_era = era_id + auction_delay; // as before
  // Update seigniorage recipients for current era
  // ... 
  let reicipients = { todo!("calculate recipients as before") };
  self.write_era_validators(delayed_era, recipients)?;
```

So my thinking is before we had to use "SeigniorageRecipientsSnapshot" container which mapped EraId into SeigniorageRecipients. Then read it from global state from single key, then extend it with new seigniorage recipients, and then trim it to contain last N eras only and write it all again under the same key. With this change we could just write new `recipients` into global state under `Key::EraValidators(delayed_era)` without reading past data. Wouldn't that work?

Reply:
> They key piece of @mpapierski's critique is correct: we need to limit the number of keys we read so that we aren't returning all the eras, just the "window" that was there before. The window's size is essentially the auction delay IIRC.

Wanted to clarify that we are talking about the _last_ N eras here with this window, so the highest `EraId`s.

Reply:
> https://github.com/darthsiroftardis/casper-node/blob/117e71b3c81c017f9b83c816069297d90b4b066a/types/src/system/auction/mod.rs#L425-L444 this code could probable become just:
> 
> ```rust
>   let delayed_era = era_id + auction_delay; // as before
>   // Update seigniorage recipients for current era
>   // ... 
>   let reicipients = { todo!("calculate recipients as before") };
>   self.write_era_validators(delayed_era, recipients)?;
> ```
> 
> So my thinking is before we had to use "SeigniorageRecipientsSnapshot" container which mapped EraId into SeigniorageRecipients. Then read it from global state from single key, then extend it with new seigniorage recipients, and then trim it to contain last N eras only and write it all again under the same key. With this change we could just write new `recipients` into global state under `Key::EraValidators(delayed_era)` without reading past data. Wouldn't that work?

Ah yes, absolutely!  I had missed the fact that @darthsiroftardis wasn't doing that because it didn't show up in the diff, so thanks for pointing this out.  

@darthsiroftardis we don't need to read the whole map and write it again in the code that @mpapierski links to.  Simply write the new recipients at `delayed_era` as he suggests.